lib/charms/opensearch/v0/opensearch_locking.py

```diff
@@ -240,7 +240,10 @@ def acquired(self) -> bool:  # noqa: C901
                 unit = self._unit_with_lock(host)
             except OpenSearchHttpError:
                 logger.exception("Error checking which unit has OpenSearch lock")
-                return False
+                # if the node lock cannot be acquired, fall back to peer databag lock
+                # this avoids hitting deadlock situations in cases where
+                # the .charm_node_lock index is not available
+                return self._peer.acquired
```

Comment:
The `self._peer.acquired` is a last-resort. We should not be using it just because we had an HTTP error. It is supposed to be used on special cases only, e.g. we have 1 or 2 units only. I think it does fit your case, as you are destroying the application.

In reality, we should use it if unit count in peer relation is <=2.

Reply:
Why should it be `<=2` and not `<=1`?

Reply:
As we've discussed, 2-node deployment does not apply the automated role management feature.

Reply:
For completeness: we no longer do automated role management. To clarify further, `<= 2` would have been a temporary workaround for the edge case where a cluster finds itself with only 2 nodes both of which having the `cluster_manager` role. One of these nodes should be added to the voting exclusion list in order to guarantee the possibility of an election in the event where something takes them both offline at the same time.

I prefer that we keep `<= 1` and address the voting exclusion separately. We're more likely to face the locking problems due to `<= 2` in the day to day development than the voting exclusion issue.

Let's please create an issue for the voting exclusion.

Reply:
<= 1 is updated, issue is created here: https://github.com/canonical/opensearch-operator/issues/327